src/class/proto_methods.rs

```diff
@@ -1,78 +1,135 @@
 use crate::ffi;
-#[cfg(not(Py_LIMITED_API))]
-use crate::ffi::PyBufferProcs;
-
-/// ABI3 doesn't have buffer APIs, so here we define the empty one.
-#[cfg(Py_LIMITED_API)]
-#[doc(hidden)]
-#[derive(Clone)]
-pub struct PyBufferProcs;
+use std::marker::PhantomData;
 
 // Note(kngwyu): default implementations are for rust-numpy. Please don't remove them.
 pub trait PyProtoMethods {
-    fn get_type_slots() -> Vec<ffi::PyType_Slot> {
-        vec![]
-    }
-    fn get_buffer() -> Option<PyBufferProcs> {
+    fn for_each_proto_slot<Visitor: FnMut(ffi::PyType_Slot)>(_visitor: Visitor) {}
+    fn get_buffer() -> Option<&'static PyBufferProcs> {
         None
     }
 }
 
-/// Typed version of `ffi::PyType_Slot`
-#[doc(hidden)]
-pub struct TypedSlot<T: Sized>(pub std::os::raw::c_int, pub T);
+pub struct PyClassProtocols<T>(PhantomData<T>);
+
+impl<T> PyClassProtocols<T> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T> Default for PyClassProtocols<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Clone for PyClassProtocols<T> {
+    fn clone(&self) -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Copy for PyClassProtocols<T> {}
 
-#[doc(hidden)]
-pub enum PyProtoMethodDef {
-    Slots(Vec<ffi::PyType_Slot>),
-    Buffer(PyBufferProcs),
+// All traits describing slots, as well as the fallback implementations for unimplemented protos
+//
+// Protos which are implented use dtolnay specialization to implement for PyClassProtocols<T>
```

Comment:
How about adding a link to https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md?
I hope our codebase would be kind for new contributors.

Reply:
👍 good idea!